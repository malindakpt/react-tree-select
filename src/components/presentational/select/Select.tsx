import { FC } from 'react';
import MinusIcon from '../../../icons/minus.svg';
import PlusIcon from '../../../icons/plus.svg';
import classes from './Select.module.scss';
import { Node } from '../../../entities/Node';

interface Props {
    node: Node;
    onLabelClick: () => void;
    onSelectionChange: (isSelected: boolean, node: Node) => void;
}
export const Select: FC<Props> = ({
    node,
    onLabelClick,
    onSelectionChange,
}: Props) => {
    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        onSelectionChange(event.target.checked, node);
    };

    return (
        <div className={classes.container}>
            <input type="checkbox" id="vehicle1" name="vehicle1" value="Bike" checked={node.isSelected} onChange={handleChange}></input>
            <div id="label" className={classes.label} onClick={() => onLabelClick()}>
                <span id="name">{node.label}</span>
                <span id="count" className={classes.count}>
                    ({node.children?.length})
                </span>
                {node.children?.length > 0 && (
                    <>{node.isExpanded ? <img width={12} src={MinusIcon} /> : <img width={12} src={PlusIcon} />}</>
                )}
            </div>
        </div>
    );
};
