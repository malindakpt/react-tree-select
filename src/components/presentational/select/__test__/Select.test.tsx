import { Select } from '../Select';
import { shallow } from 'enzyme';

describe('Select Component', () => {
    it('renders elements without crashing', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable
                isExpanded
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );

        const checkbox = wrapper.find('#checkbox');
        const label = wrapper.find('#label');
        const name = wrapper.find('#name');
        const count = wrapper.find('#count');
        const upArrow = wrapper.find('#upArrow');
        const rightArrow = wrapper.find('#rightArrow');

        expect(checkbox).toHaveLength(1);
        expect(label).toHaveLength(1);
        expect(name).toHaveLength(1);
        expect(count).toHaveLength(1);

        expect(upArrow).toHaveLength(1);
        expect(rightArrow).toHaveLength(0);

        expect(wrapper).toMatchSnapshot();
    });

    it('renders name span with the correct name', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable
                isExpanded={false}
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );
        const name = wrapper.find('#name');

        expect(name.text()).toEqual('label1');
        expect(wrapper).toMatchSnapshot();
    });

    it('renders count span with the brackets', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable
                isExpanded={false}
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );
        const count = wrapper.find('#count');

        expect(count.text()).toEqual('(10)');
        expect(wrapper).toMatchSnapshot();
    });

    it('renders rightArrow when not expanded', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable
                isExpanded={false}
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );
        const rightArrow = wrapper.find('#rightArrow');
        const upArrow = wrapper.find('#upArrow');

        expect(rightArrow).toHaveLength(1);
        expect(upArrow).toHaveLength(0);

        expect(wrapper).toMatchSnapshot();
    });

    it('renders non of the arrows when Select is not expandable', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable={false}
                isExpanded
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );
        const rightArrow = wrapper.find('#rightArrow');
        const upArrow = wrapper.find('#upArrow');

        expect(rightArrow).toHaveLength(0);
        expect(upArrow).toHaveLength(0);

        expect(wrapper).toMatchSnapshot();
    });

    it('onLabelClick callback on label click', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable
                isExpanded
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );

        const label = wrapper.find('#label');
        label.simulate('click');
        expect(onLabelClick).toHaveBeenCalled();

        expect(wrapper).toMatchSnapshot();
    });

    it('onSelectionChange callback on checkbox change', () => {
        const onSelectionChange = jest.fn();
        const onLabelClick = jest.fn();
        const wrapper = shallow(
            <Select
                name="label1"
                count={10}
                isExpandable
                isExpanded
                isSelected
                onSelectionChange={onSelectionChange}
                onLabelClick={onLabelClick}
            />,
        );

        const checkbox = wrapper.find('#checkbox');

        checkbox.simulate('change', { target: { checked: true } });
        expect(onSelectionChange).toHaveBeenCalledWith(true);

        checkbox.simulate('change', { target: { checked: false } });
        expect(onSelectionChange).toHaveBeenCalledWith(false);

        expect(wrapper).toMatchSnapshot();
    });
});
