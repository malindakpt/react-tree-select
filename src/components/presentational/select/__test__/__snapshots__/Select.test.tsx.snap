// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component onLabelClick callback on label click 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
    <Memo(ForwardRef(ArrowDropUpIcon))
      id="upArrow"
    />
  </div>
</div>
`;

exports[`Select Component onSelectionChange callback on checkbox change 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
    <Memo(ForwardRef(ArrowDropUpIcon))
      id="upArrow"
    />
  </div>
</div>
`;

exports[`Select Component renders count span with the brackets 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
    <Memo(ForwardRef(ArrowRightIcon))
      id="rightArrow"
    />
  </div>
</div>
`;

exports[`Select Component renders elements without crashing 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
    <Memo(ForwardRef(ArrowDropUpIcon))
      id="upArrow"
    />
  </div>
</div>
`;

exports[`Select Component renders name span with the correct name 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
    <Memo(ForwardRef(ArrowRightIcon))
      id="rightArrow"
    />
  </div>
</div>
`;

exports[`Select Component renders non of the arrows when Select is not expandable 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
  </div>
</div>
`;

exports[`Select Component renders rightArrow when not expanded 1`] = `
<div
  className="container"
>
  <WithStyles(ForwardRef(Checkbox))
    checked={true}
    color="primary"
    id="checkbox"
    onChange={[Function]}
  />
  <div
    className="label"
    id="label"
    onClick={[Function]}
  >
    <span
      id="name"
    >
      label1
    </span>
    <span
      className="count"
      id="count"
    >
      (
      10
      )
    </span>
    <Memo(ForwardRef(ArrowRightIcon))
      id="rightArrow"
    />
  </div>
</div>
`;
